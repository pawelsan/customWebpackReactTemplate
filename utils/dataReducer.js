export const dataReducer = (state, action) => {
  if (action.type === "SET_ERROR") {
    return { ...state, list: [], error: true };
  }

  if (action.type === "SET_LIST") {
    return { ...state, list: action.list, error: null };
  }

  throw new Error();
};
