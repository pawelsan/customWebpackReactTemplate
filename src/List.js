import React, { useReducer, useEffect } from "react";
import axios from "axios";
import { dataReducer } from "../utils/dataReducer";

const List = () => {
  const initialData = {
    list: [],
    error: null,
  };

  const [data, dispatch] = useReducer(dataReducer, initialData);

  useEffect(() => {
    axios
      .get("http://hn.algolia.com/api/v1/search?query=react")
      .then((response) => {
        dispatch({ type: "SET_LIST", list: response.data.hits });
      })
      .catch(() => {
        dispatch({ type: "SET_ERROR" });
      });
  }, []);
  return (
    <div>
      <h2>My Async Data</h2>

      {data.error && <div className="error">Error</div>}

      <ul>
        {data.list.map((item) => (
          <li key={item.objectID}>{item.title}</li>
        ))}
      </ul>
    </div>
  );
};

export default List;
