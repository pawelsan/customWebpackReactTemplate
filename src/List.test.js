import React from "react";
import renderer from "react-test-renderer";

import List from "./List";
import { dataReducer } from "../utils/dataReducer";

const list = ["a", "b", "c"];

describe("List", () => {
  test("snapshot renders", () => {
    const component = renderer.create(<List />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe("Reducer", () => {
    it("should set a list", () => {
      const state = { list: [], error: null };
      const newState = dataReducer(state, {
        type: "SET_LIST",
        list,
      });

      expect(newState).toEqual({ list, error: null });
    });
    it("should reset the error if list is set", () => {
      const state = { list: [], error: true };
      const newState = dataReducer(state, {
        type: "SET_LIST",
        list,
      });

      expect(newState).toEqual({ list, error: null });
    });

    it("should set the error", () => {
      const state = { list: [], error: null };
      const newState = dataReducer(state, {
        type: "SET_ERROR",
      });

      expect(newState.error).toBeTruthy();
    });
  });
});
