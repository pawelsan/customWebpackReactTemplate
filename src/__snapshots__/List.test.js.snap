// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List snapshot renders 1`] = `
<div>
  <h2>
    My Async Data
  </h2>
  <ul />
</div>
`;
